op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -515,116 +517,47 @@ func runDBCheck(opts dbCheckOptions) (err error) {
+	var errResult *multierror.Error
+	var errMu sync.Mutex
+	g := new(errgroup.Group)
+	g.SetLimit(runtime.NumCPU() * 2)
+	lastBlockNumber := *rawdb.ReadHeaderNumber(nonAncientDB, rawdb.ReadHeadHeaderHash(nonAncientDB))
+	for i := uint64(0); i <= lastBlockNumber; i += opts.batchSize {
+		start := i
 		g.Go(func() error {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			default:
-				blockRange, err := loadRangeFunc(start, count)
-				if err != nil {
-					err = fmt.Errorf("failed to load block range: %w", err)
-					if opts.failFast {
-						return err
-					}
-					log.Error(err.Error())
-					mu.Lock()
-					errResult = multierror.Append(errResult, err)
-					mu.Unlock()
-					return nil
-				}
-				if _, err := blockRange.CheckContinuity(nil, count); err != nil {
-					err = fmt.Errorf("failed continuity check: %w", err)
-					if opts.failFast {
-						return err
-					}
-					log.Error(err.Error())
-					mu.Lock()
-					errResult = multierror.Append(errResult, err)
-					mu.Unlock()
-					return nil
+			end := min(start+opts.batchSize-1, lastBlockNumber)
+			if start > 0 {
+				start -= 1 // Ensure we check continuity across batch boundaries
+			}
+			log.Info("Checking data continuity for block range", "start", start, "end", end, "count", end-start+1)
+			err := checkRange(ancientDB, nonAncientDB, start, end)
+			if err != nil {
+				err = fmt.Errorf("failed to check range (%d-%d): %w", start, end, err)
+				if opts.failFast {
+					return err
 				}
-				log.Info("Successfully checked block range continuity", "start", start, "end", start+count-1, "count", count)
-				return nil
+				errMu.Lock()
+				defer errMu.Unlock()
+				errResult = multierror.Append(errResult, err)
 			}
+			return nil
 		})
 	}
-	checkContinuity := func(start, end uint64, loadRangeFunc func(uint64, uint64) (*RLPBlockRange, error)) error {
-		if (start <= lastAncientNumber && end > lastAncientNumber) || (end > lastBlockNumber) || (end < start) {
-			return fmt.Errorf("invalid range for continuity check: start=%d, end=%d, lastAncientNumber=%d, lastBlockNumber=%d", start, end, lastAncientNumber, lastBlockNumber)
-		}
-		for i := start; i <= end; i += opts.batchSize {
-			count := min(opts.batchSize, end-i+1)
-			checkRange(i, count, loadRangeFunc)
-		}
-		return nil
-	}
-
-	log.Info("Checking continuity of ancient blocks", "start", 0, "end", lastAncientNumber, "count", lastAncientNumber+1)
-	if err := checkContinuity(0, lastAncientNumber, func(start, count uint64) (*RLPBlockRange, error) {
-		return loadAncientRange(ancientDB, start, count)
-	}); err != nil {
-		return err
-	}
-	log.Info("Checking continuity of non-ancient blocks", "start", lastAncientNumber+1, "end", lastBlockNumber, "count", lastBlockNumber-lastAncientNumber)
-	if err := checkContinuity(lastAncientNumber+1, lastBlockNumber, func(start, count uint64) (*RLPBlockRange, error) {
-		return loadNonAncientRange(nonAncientDB, start, count)
-	}); err != nil {
+	err = g.Wait()
+	if opts.failFast {
```

Comment:
I don't think we need to check `failFast` here. If `g.Wait` returns an error it's either because we already checked `failFast` inside a go routine or because something unexpected happened and an error was thrown by something other than `checkRange`